rfcs/0145-doc-strings.md

```diff
@@ -0,0 +1,501 @@
+
+While this allows the most freedom, it is usually considered the best option, not creating any restrictions. 
+
+However, nix/rfc72 defines commonMark as the official documentation format. 
+This is why we decided to follow this convention.
+
+## Consequences of not implementing this
+
+- By not implementing this feature, Nix gains no ability for tool-generated documentation.
+- Documentation will be defined by nixdoc, not by the nix community.
+- Many existing comments written for documentation will remain imperceptible.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+- Will `nix` itself implement native support like in rust -> `cargo doc`?
+
+- How can a tool keep the connection from where a docstring was defined and where the attribute was exposed (lib/default. Nix exposes mapAttrs which is defined at lib/attrsets. Nix)
+  - There are more complicated things.
+
+**Possible answer**: This is a tooling question; implementation details will not be discussed in detail. It is possible to track comments in the AST. As shown before in this PR: [nix/#1652](https://github.com/NixOS/nix/pull/1652)
+
+# Future work
+[Future]: #future-work
+
+## Editor support
+
+- Implement displaying the related documentation when hovering over an expression. (lspAction/hover)
+
+Nix already offers a bunch of LSPs, e.g., [nil](https://github.com/oxalica/nil), [rnix-lsp](https://github.com/nix-community/rnix-lsp) are the most common ones.
+
+
+## Nixdoc
+
+Nixdoc must be changed to differentiate between regular comments and doc-comments.
+There might be an intermediate phase of transition, where the old Syntax and features are supported in parallel to allow a phase of transition and refactoring of existing documentation comments.
+
+## Documentation generators
+
+- Future nixdoc could have a static `map.json` that contains the scope for every discovery file/path in nixpkgs.
+
+As this second approach is much easier, we propose that we initially start to extend the scope.
+
+Generating static documentation from doc-comments would still be excellent but remains a challenge if generated from Nix dynamically.
+If we want the fully automated approach, we need something to evaluate nix expressions. 
+
+We solve such concerns in `tvix` or in `nix`, which could provide a `nix-analyzer` that pre-evaluates expressions and their respective documentation. 
+
+## Type
+
+- An RFC under construction specifies the used Syntax within the `# Type` Heading.
+- The `type` feature should belong in the nix syntax. Try them within the comments first; This is still possible.
+
+- see a [preview](https://typednix.dev) of an eventual future doc-type-syntax.
+
+## Native support in Nix
+
+- `NixOS/nix` should implement native support for doc-comments so that our users do not have to rely on nixpkgs or external tools. Those tools can still exist and provide more custom functionality, but documenting nix expressions should be natively possible.
```

Comment:
I suspect something else meant under word "native".

What does "native" support and "documenting natively" means?  
Is it that Nix should allow to introspecting doc-comments right in Nix expressions themselves?

It also worth to relate this with some goal of this RFC.

E.g. if it is about not having `nix` and `nix-doc` in separate packages, then I don't think it is much useful for this RFC 🙂

Reply:
As this is in the "Future work" section, this was meant to show a future possibility.

> If it is about not having nix and nix-doc in separate packages ... 

Yes. It is about having a "nixdoc" like functionality directly in nix. (not separate packages, if you will)

How would you re-write this sentence to make that more clear?

Reply:
With mention of "Future work", I think I understood you now. It is about how we can leverage results of this RFC which is indeed useful for its adoption. From that I only can think of:

```markdown
- `nix repl` could fallback to use doc-comments for output of `:doc`.
- `nix repl` could enrich `:t` (describing result of evaluation) to include type signature.
  E.g. instead of plain "a function" could say "a function of Any -> integer".
- `nix --show-trace` can be extended to include type information in output when available.
- `nix` could be extended with `--check-types` option to print warnings/errors when from doc-comments 
  it is known that types do not match either for documented expression or on the boundaries of calls like
  when passing in function.
```

I can see why "nixdoc" (I'm not familiar with it yet) might be useful in `nix repl`. But I do not see at the moment why it is important to have it in a same package and I do not see concerns with using "external" tools. E.g.
What us impact on this RFC we will have by not following that?  
What impact this RFC have on decision to follow that suggestion?

Note that if it is about ensuring we have coherent parsing in `nix` and in tools around generating documentation - it can be solved by having AST parser vended from Nix or have it in a separate package consumed by all tools including LSP.

Reply:
Thanks for your pretty detailed response.

I think I won't go into that much detail in this section. That all is what I meant with the term "native support"

> Ensuring we have coherent parsing in Nix

That'd be nice but it's not the target of this RFC.

Also, the whole point is not about moving any "package" into one another or something. Just that nix could natively implement (same or more)  functionality that we currently have by calling out to nixdoc